### PR TITLE
Develop to Master:  v0.11.1

### PR DESCRIPTION
**Pull request by profxj**
Now the big push over.  To remind us, here are all the CHANGES in v0.11.1:

- Add VLT/FORS filters to our database
- Improved DEIMOS frame typing
- Brings Gemini/GMOS into the suite (R400)
- Also an important change for autoid.full_template()
- Fixed trace extrapolation, to fix bugs in object finding. Tweaks to
  object finding algorithm.
- Major improvements to echelle object finding.
- Improved outlier rejection and coefficient fitting in pca_trace
- Major improvements to coadd routines in coadd1d
- Introduced telluric module and telluric correction routines
- Implemented tilt image type which is now a required frame type
- Streamlined and abstracted echelle properties and echelle routine in
  spectrograph classes.
- Revamped 2-d coadding routines and introduced 2-d coadding of
  MultiSlit data
- Improved ginga plotting routines.
- Fixed bug associated with astropy.stats.sigma_clipped_stats when
  astropy.stats.mad_std is used.
- Refactor BPM generation
- Merge raw_image loading with datasec_img and oscansec_img generation
- Sync datasec_img to image in ProcessRawImage
- Started (barely) on a path to having calibration images in counts and
  not ADU
- Refactors GMOS for get_rawimage method
- Enables GMOS overscan subtraction
- Adds R400 wavelength solution for old E2V chip
- Revises simple_calib() method for quick and dirty wavelength
  calibration
- Adds a related show_wvcalib script
- Changes to ech_combspec to better treat filenames
- Fixed bug when bias was set to 'force' which was not bias subtracting
- Implemented changes to vlt_xshooter_nir to now require darks taken
  between flats
- Made flat fielding code a bit more robust against hot pixels at edge
  of orders
- Added pypeit_chk_flat script to view flat images
- Refactored image objects into RawImage, ProcessRawImage, PypeItImage,
  BuildImage
- Moved load() and save() methods from MasterFrame to the individual
  calibration objects
- Converted ArcImage and FlatImages into counts
- Added code to allow for IVAR and RN2 image generation for calibs
- Added several from_master_file() instantiation methods
- Use coadd2d.weighted_combine() to stack calibration images
- Major refactor of slit edge tracing
- Deprecated previous tracing code: `pypeit.traceslits` and
  `pypeit.core.trace_slits`, as well as some functions in
  `pypeit.core.extract` that were replaced by
  `pypeit.core.moment.moment1d` and functions in `pypeit.core.trace`.
- PCA now saved to MasterEdges file; added I/O methods

### CODE REVIEW

File: pypeit/core/flat.py [hunk 1]
```diff
@@ -818,232 +846,232 @@ def norm_slits(mstrace, datasec_img, lordloc, rordloc, pixwid,
...
-    # Only do a PCA if there are enough good orders
-    if np.sum(1.0 - extrap_slit) > sofit[0] + 1:
-        # Perform a PCA on the tilts
-        msgs.info("Performing a PCA on the spatial slit profiles")
-        ordsnd = np.arange(nslits) + 1.0
-        xcen = spatfit[:, np.newaxis].repeat(nslits, axis=1)
-        fitted, outpar = pca.basis(xcen, sltval, fitcoeff, lnpc, sofit, x0in=ordsnd, mask=maskord, skipx0=False,
-                                     function=fitfunc)
-        if not debug:
+#def slit_profile_pca(mstrace, tilts, msblaze, extrap_slit, slit_profiles,
```
Comment (profxj) [inline pypeit/core/flat.py]: Remove

File: pypeit/core/load.py [hunk 2]
```diff
@@ -150,7 +156,193 @@ def load_specobjs(fname,order=None):
     # Return
     return sobjs, head0
 
-def load_spec_order(fname,objid=None,order=None,extract='OPT',flux=True):
+# TODO I don't think we need this routine
+def load_ext_to_array(hdulist, ext_id, ex_value='OPT', flux_value=True, nmaskedge=None):
```
Comment (profxj) [inline pypeit/core/load.py]: I am not going to fuss about these methods as most will disappear soon.

File: pypeit/core/procimg.py [hunk 3]
```diff
@@ -279,22 +248,20 @@ def gain_frame(amp_img, gain, trim=True):
...
             amplifier 1 is provided by `gain[0]`, etc.
-        trim (:obj:`bool`, optional):
-            Trim the overscan section from the image.
 
     Returns:
         `numpy.ndarray`: Image with the gain for each pixel.
     """
     # TODO: Remove this or actually do it.
-    msgs.warn("Should probably be measuring the gain across the amplifier boundary")
+    # msgs.warn("Should probably be measuring the gain across the amplifier boundary")
```
Comment (profxj) [inline pypeit/core/procimg.py]: Remove

File: pypeit/core/save.py [hunk 4]
```diff
@@ -216,6 +217,82 @@ def save_1d_spectra_fits(specObjs, header, spectrograph, outfile, helio_dict=Non
     return outfile
 
 
+def save_coadd1d_to_fits(outfile, waves, fluxes, ivars, masks, telluric=None, obj_model=None,
```
Comment (profxj) [inline pypeit/core/save.py]: Same as above for `load`.  Almost all of these methods will disappear
soon, i.e. in v0.11.2

File: pypeit/core/skysub.py [hunk 5]
```diff
@@ -40,22 +44,20 @@ def skysub_npoly(thismask):
...
-
 def global_skysub(image, ivar, tilts, thismask, slit_left, slit_righ, inmask = None, bsp=0.6, sigrej=3.0, maxiter=35,
                   trim_edg=(3,3), pos_mask=True, show_fit=False, no_poly=False, npoly=None):
     """
     Perform global sky subtraction on an input slit
 
-    Parameters
-    ----------
+    Args:
     image: float ndarray, shape (nspec, nspat)
```
Comment (profxj) [inline pypeit/core/skysub.py]: Sphinx

File: pypeit/core/telluric.py [hunk 6]
```diff
@@ -0,0 +1,1259 @@
+
```
Comment (profxj) [inline pypeit/core/telluric.py]: Module needs a brief description.

File: pypeit/core/telluric.py [hunk 7]
```diff
@@ -0,0 +1,1259 @@
...
+from IPython import embed
+from pypeit.spectrographs.util import load_spectrograph
+
+
+##############################
+#  Telluric model functions  #
+##############################
+
+#def get_bounds_tell(tell_dict, resln_guess, resln_frac_bounds, pix_shift_bounds):
```
Comment (profxj) [inline pypeit/core/telluric.py]: remove this commented code

File: pypeit/core/telluric.py [hunk 8]
```diff
@@ -0,0 +1,1259 @@
...
+#    # Guess resolution from wavelength sampling of telluric grid if it is not provided
+#    resln_guess = 1.0 / (3.0 * dloglam * np.log(10.0))  # assume roughly Nyquist sampling
+#    pix_per_sigma = 1.0 / resln_guess / (dloglam * np.log(10.0)) / (2.0 * np.sqrt(2.0 * np.log(2)))
+#
+#    return dloglam, resln_guess, pix_per_sigma
+
+
+# TODO These codes should probably be in a separate qso_pca module. Also pickle functionality needs to be removed.
+def init_pca(filename,wave_grid,redshift, npca):
+    # Read in the pickle file from coarse_pca.create_coarse_pca
```
Comment (profxj) [inline pypeit/core/telluric.py]: Needs a doc string

File: pypeit/core/telluric.py [hunk 9]
```diff
@@ -0,0 +1,1259 @@
...
+    pca_comp_new = pca_interp(wave_grid)
+    # Generate a mixture model for the coefficients prior, what should ngauss be?
+    prior = mixture.GaussianMixture(n_components = npca-1).fit(coeffs_c[:, 1:npca])
+    # Construct the PCA dict
+    pca_dict = {'npca': npca, 'components': pca_comp_new, 'prior': prior, 'coeffs': coeffs_c,
+                'z_fid': redshift, 'dloglam': dloglam}
+    return pca_dict
+
+def pca_eval(theta,pca_dict):
+    # theta_pca[0] is redshift
```
Comment (profxj) [inline pypeit/core/telluric.py]: doc string.

File: pypeit/core/telluric.py [hunk 10]
```diff
@@ -0,0 +1,1259 @@
...
+    tell_pad_pix = int(np.ceil(10.0 * pix_per_sigma))
+
+    tell_dict = dict(wave_grid=wave_grid, dloglam=dloglam,
+                     resln_guess=resln_guess, pix_per_sigma=pix_per_sigma, tell_pad_pix=tell_pad_pix,
+                     pressure_grid=pg, temp_grid=tg, h2o_grid=hg, airmass_grid=ag, tell_grid=model_grid)
+    return tell_dict
+
+
+def interp_telluric_grid(theta,tell_dict):
+
```
Comment (profxj) [inline pypeit/core/telluric.py]: doc string.  I'll stop repeating it.  Obviously
this is pervasive.  Maybe we should remove this
module altogether.

File: pypeit/core/telluric.py [hunk 11]
```diff
@@ -0,0 +1,1259 @@
...
+# Polynomial evaluation function.
+def eval_poly_model(theta, obj_dict):
+    return obj_dict['polymodel'], (obj_dict['polymodel'] > 0.0)
+
+
+# User defined functions
+# obj_dict, bounds_obj = init_obj_model(obj_params, iord, wave, flux, ivar, mask, tellmodel)
+# obj_model, modelmask =  eval_obj_model(theta_obj, obj_dict)
+
+class Telluric(object):
```
Comment (profxj) [inline pypeit/core/telluric.py]: This class should not be in core.

File: pypeit/core/tracewave.py [hunk 12]
```diff
@@ -671,6 +702,8 @@ def fit2tilts(shape, coeff2, func2d):
         result of griddata tilt fit
     func2d: str
```
Comment (profxj) [inline pypeit/core/tracewave.py]: Sphinx (and use `Args` not `Parameters`)

File: pypeit/core/wave.py [hunk 13]
```diff
@@ -52,7 +51,10 @@ def flex_shift(obj_skyspec, arx_skyspec, mxshft=20):
     flex_dict: dict
       Contains flexure info
     """
-    flex_dict = {}
+
+    # TODO None of these routines should have dependencies on XSpectrum1d!
```
Comment (profxj) [inline pypeit/core/wave.py]: Do any?

File: pypeit/deprecated/combine.py [hunk 14]
```diff
@@ -0,0 +1,204 @@
+
+# TODO JFH This code is a giant piece of crapp that should be rewritten from scratch
```
Comment (profxj) [inline pypeit/deprecated/combine.py]: JFH should remove this insulting TODO, especially since
it has been deprecated.

File: pypeit/sampling.py [hunk 15]
```diff
@@ -0,0 +1,728 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
```
Comment (profxj) [inline pypeit/sampling.py]: Should this module be in core?

**Comment by jhennawi** [comment]
Closing until this mess with edge_merge PR is fixed. This should be resubmitted.

**Pull request closed without merging**